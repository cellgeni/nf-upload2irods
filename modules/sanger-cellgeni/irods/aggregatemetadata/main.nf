process IRODS_AGGREGATEMETADATA {
    tag "Aggregating metadata for ${meta.id}"
    container 'docker://quay.io/cellgeni/metacells-python:latest'

    input:
    tuple val(meta), path(irods_metadata, name: "input.csv")

    output:
    tuple val(meta), path("metadata.csv"), emit: csv
    tuple val(meta), path("metadata.json"), emit: json
    path "versions.yml", emit: versions

    script:
    def args = task.ext.args ?: '--dup-sep ";" --index_name "id"'
    """
    aggregate_metadata.py \
        ${args} \
        --input input.csv \
        --id ${meta.id}
        

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python --version | awk '{print \$2}')
        pandas: \$( python -c "import pandas; print(pandas.__version__)" )
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: '--dup-sep "," --index_name "id"'
    """
    touch metadata.csv metadata.json
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python --version | awk '{print \$2}')
        pandas: \$( python -c "import pandas; print(pandas.__version__)" )
    END_VERSIONS
    """
}
