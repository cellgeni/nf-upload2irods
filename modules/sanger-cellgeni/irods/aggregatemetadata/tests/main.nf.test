nextflow_process {
    name "Test Process: IRODS_AGGREGATEMETADATA"
    script "../main.nf"
    process "IRODS_AGGREGATEMETADATA"

    tag "modules"
    tag "irods"
    tag "irods/aggregatemetadata"
    tag "modules_sangercellgeni"

    setup {
        run("IRODS_GETMETADATA", alias: "CRAM") {
            script "../../getmetadata/main.nf"
            config "../../getmetadata/module.config"
            process {
                """
                input[0] = [ [id: "cram"], "/seq/illumina/runs/41/41796/lane1/plex1/41796_1#1.cram" ]
                """
            }
        }

        run("IRODS_GETMETADATA", alias: "CELLRANGER_ARC_OUTPUT") {
            script "../../getmetadata/main.nf"
            config "../../getmetadata/module.config"
            process {
                """
                input[0] = [ [id: "cellranger_arc"], "/seq/illumina/cellranger-arc/cellranger-arc202_count_43c2d8dd1eaf98b635896165fd98ae3a" ]
                """
            }
        }

        run("IRODS_GETMETADATA", alias: "EMPTY") {
            script "../../getmetadata/main.nf"
            config "../../getmetadata/module.config"
            process {
                """
                input[0] = [ [id: "empty"], "/archive/cellgeni/multiome" ]
                """
            }
        }
    }

    test("Cellranger ARC output metadata aggregation") {
        tag "cellranger_arc"
        config "../module.config"

        when {
            process {
                """
                input[0] = CELLRANGER_ARC_OUTPUT.out.csv
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.versions },
                { assert process.out.csv },
                { assert process.out.json },
                { assert path(process.out.csv.get(0).get(1)).csv.rowCount == 1 },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test(".cram file metadata aggregation") {
        tag "cram"
        config "../module.config"

        when {
            process {
                """
                input[0] = CRAM.out.csv
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.versions },
                { assert process.out.csv },
                { assert process.out.json },
                { assert path(process.out.csv.get(0).get(1)).csv.rowCount == 1 },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Empty file output metadata aggregation") {
        tag "empty"
        config "../module.config"

        when {
            process {
                """
                input[0] = EMPTY.out.csv
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.versions },
                { assert process.out.csv },
                { assert process.out.json },
                { assert path(process.out.csv.get(0).get(1)).csv.rows[0] == ["irodspath": "empty"] },
                { assert snapshot(process.out).match() }
            )
        }
    }
}