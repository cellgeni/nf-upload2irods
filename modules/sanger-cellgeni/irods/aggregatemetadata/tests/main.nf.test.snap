{
    "Empty file output metadata aggregation": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "empty"
                        },
                        "metadata.csv:md5,cbac237f9f819fadf37c85efd0b4269e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "empty"
                        },
                        "metadata.json:md5,22e67cc3ae278cb47bca0058382d3330"
                    ]
                ],
                "2": [
                    "versions.yml:md5,c7179ef11646eed9f4262bc31b48d134"
                ],
                "csv": [
                    [
                        {
                            "id": "empty"
                        },
                        "metadata.csv:md5,cbac237f9f819fadf37c85efd0b4269e"
                    ]
                ],
                "json": [
                    [
                        {
                            "id": "empty"
                        },
                        "metadata.json:md5,22e67cc3ae278cb47bca0058382d3330"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c7179ef11646eed9f4262bc31b48d134"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-10-24T16:32:25.841505619"
    },
    ".cram file metadata aggregation": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "cram"
                        },
                        "metadata.csv:md5,6ea84f3640a255d1adf9578b6f0c35f4"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "cram"
                        },
                        "metadata.json:md5,a5eac87d4b5894c5dacda14d5261cb3e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,c7179ef11646eed9f4262bc31b48d134"
                ],
                "csv": [
                    [
                        {
                            "id": "cram"
                        },
                        "metadata.csv:md5,6ea84f3640a255d1adf9578b6f0c35f4"
                    ]
                ],
                "json": [
                    [
                        {
                            "id": "cram"
                        },
                        "metadata.json:md5,a5eac87d4b5894c5dacda14d5261cb3e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c7179ef11646eed9f4262bc31b48d134"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-10-24T16:40:51.463429041"
    },
    "Cellranger ARC output metadata aggregation": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "cellranger_arc"
                        },
                        "metadata.csv:md5,22087d13d841154d0a548f161fd8e723"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "cellranger_arc"
                        },
                        "metadata.json:md5,c6f62aeba959b75ab8e476fe6c18eb26"
                    ]
                ],
                "2": [
                    "versions.yml:md5,c7179ef11646eed9f4262bc31b48d134"
                ],
                "csv": [
                    [
                        {
                            "id": "cellranger_arc"
                        },
                        "metadata.csv:md5,22087d13d841154d0a548f161fd8e723"
                    ]
                ],
                "json": [
                    [
                        {
                            "id": "cellranger_arc"
                        },
                        "metadata.json:md5,c6f62aeba959b75ab8e476fe6c18eb26"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c7179ef11646eed9f4262bc31b48d134"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-10-24T16:40:01.098156284"
    }
}